File: training-data/go/min_stack.go
package main

import "fmt"

type minStack struct {
	items, mins []int
}

func (s *minStack) push(value int) {
	s.items = append(s.items, value)
	if len(s.mins) == 0 || value < s.mins[len(s.mins)-1] {
		s.mins = append(s.mins, value)
	} else {
		s.mins = append(s.mins, s.mins[len(s.mins)-1])
	}
}

func (s *minStack) pop() int {
	n := len(s.items) - 1
	item := s.items[n]
	s.items = s.items[:n]
	s.mins = s.mins[:n]
	return item
}

func (s *minStack) getMin() int {
	return s.mins[len(s.mins)-1]
}

func main() {
	s := &minStack{}
	s.push(3)
	s.push(1)
	s.push(2)
	fmt.Println(s.getMin())
	s.pop()
	fmt.Println(s.getMin())
}

File: training-data/rust/min_stack.rs
struct MinStack {
    items: Vec<i32>,
    mins: Vec<i32>,
}

impl MinStack {
    fn new() -> Self {
        MinStack { items: vec![], mins: vec![] }
    }

    fn push(&mut self, value: i32) {
        self.items.push(value);
        let current_min = match self.mins.last() {
            Some(&m) => value.min(m),
            None => value,
        };
        self.mins.push(current_min);
    }

    fn pop(&mut self) -> Option<i32> {
        self.mins.pop();
        self.items.pop()
    }

    fn get_min(&self) -> Option<i32> {
        self.mins.last().copied()
    }
}

fn main() {
    let mut s = MinStack::new();
    s.push(3);
    s.push(1);
    s.push(2);
    println!("{:?}", s.get_min());
    s.pop();
    println!("{:?}", s.get_min());
}
